tools/dotnet-linker/Steps/ManagedRegistrarStep.cs

```diff
@@ -0,0 +1,1162 @@
+				il.Append (callAllocateNSObject); // ldarg_0
+				il.Emit (OpCodes.Ldc_I4_2); // NSObject.Flags.NativeRef
+				il.Emit (OpCodes.Call, git);
+				il.Emit (OpCodes.Dup); // this is for the call to ObjCRuntime.NativeObjectExtensions::GetHandle after the call to the constructor
+			} else {
+				// instance method
+				il.Emit (OpCodes.Ldarg_0);
+				EmitConversion (method, il, method.DeclaringType, true, -1, out var nativeType, postProcessing, selfVariable);
+			}
+
+			callback.AddParameter ("sel", abr.System_IntPtr);
+
+			var managedParameterCount = 0;
+			var nativeParameterOffset = isInstanceCategory ? 1 : 2;
+			var parameterStart = isInstanceCategory ? 1 : 0;
+			if (method.HasParameters)
+				managedParameterCount = method.Parameters.Count;
+
+			if (isGeneric) {
+				il.Emit (OpCodes.Ldc_I4, managedParameterCount);
+				il.Emit (OpCodes.Newarr, abr.System_Object);
+			}
+
+			if (method.HasParameters) {
+				var isDynamicInvoke = isGeneric;
+				for (var p = parameterStart; p < managedParameterCount; p++) {
+					var nativeParameter = callback.AddParameter ($"p{p}", placeholderType);
+					var nativeParameterIndex = p + nativeParameterOffset;
+					var managedParameterType = method.Parameters [p].ParameterType;
+					var baseParameter = baseMethod.Parameters [p];
+					var isOutParameter = IsOutParameter (method, p, baseParameter);
+					if (isDynamicInvoke && !isOutParameter) {
+						if (parameterStart != 0) {
+							AddException (ErrorHelper.CreateError (99, $"Unexpected parameterStart {parameterStart} in method {GetMethodSignature (method)} for parameter {p}"));
+							continue;
+						}
+						il.Emit (OpCodes.Dup);
+						il.Emit (OpCodes.Ldc_I4, p);
+					}
+					if (!isOutParameter) {
+						il.EmitLoadArgument (nativeParameterIndex);
+					}
+					if (EmitConversion (method, il, managedParameterType, true, p, out var nativeType, postProcessing, selfVariable, isOutParameter, nativeParameterIndex, isDynamicInvoke)) {
+						nativeParameter.ParameterType = nativeType;
+					} else {
+						nativeParameter.ParameterType = placeholderType;
+						AddException (ErrorHelper.CreateError (99, "Unable to emit conversion for parameter {2} of type {0}. Method: {1}", method.Parameters [p].ParameterType, GetMethodSignatureWithSourceCode (method), p));
+					}
+					if (isDynamicInvoke && !isOutParameter) {
+						if (managedParameterType.IsValueType)
+							il.Emit (OpCodes.Box, managedParameterType);
+						il.Emit (OpCodes.Stelem_Ref);
+					}
+				}
+			}
+
+			if (callSuperParameter is not null)
+				callback.Parameters.Add (callSuperParameter);
+
+			callback.AddParameter ("exception_gchandle", new PointerType (abr.System_IntPtr));
+
+			if (isGeneric) {
+				il.Emit (OpCodes.Call, abr.MethodBase_Invoke);
+				if (isVoid) {
+					il.Emit (OpCodes.Pop);
+				} else if (method.ReturnType.IsValueType) {
+					il.Emit (OpCodes.Unbox_Any, method.ReturnType);
+				} else {
+					// il.Emit (OpCodes.Castclass, method.ReturnType);
```

Comment:
To be fair I did not manage to follow through what exactly happens at this point in the code, just wanted to check whether this comment has been left intentionally?

Reply:
@ivanpovazan I've improved this comment a bit.